### PR TITLE
[clang][Interp] Implement integral->complex casts

### PR DESCRIPTION
**Pull request by tbaederr**
Allocate storage for them,  initialize the first member with the given value and the second member to 0.

### CODE REVIEW

File: clang/lib/AST/Interp/ByteCodeExprGen.cpp [hunk 1]
```diff
@@ -283,6 +283,28 @@ bool ByteCodeExprGen<Emitter>::VisitCastExpr(const CastExpr *CE) {
   case CK_ToVoid:
     return discard(SubExpr);
 
+  case CK_IntegralRealToComplex:
+  case CK_FloatingRealToComplex: {
+    // We're creating a complex value here, so need to
```
Comment (cor3ntin) [inline clang/lib/AST/Interp/ByteCodeExprGen.cpp:288]:
```suggestion
    // We're creating a complex value here, so we need to
```

**Review by cor3ntin — APPROVED** [review]

**Comment by shafik** [comment]
Can we please have a more detailed summary for this PR? We rely on more detailed summaries in git logs.

File: clang/lib/AST/Interp/ByteCodeExprGen.cpp [hunk 2]
```diff
@@ -283,6 +283,28 @@ bool ByteCodeExprGen<Emitter>::VisitCastExpr(const CastExpr *CE) {
...
+    if (!Initializing) {
+      std::optional<unsigned> LocalIndex =
+          allocateLocal(CE, /*IsExtended=*/true);
+      if (!LocalIndex)
+        return false;
+      if (!this->emitGetPtrLocal(*LocalIndex, CE))
+        return false;
+    }
+
+    if (!this->visitArrayElemInit(0, SubExpr))
```
Comment (shafik) [inline clang/lib/AST/Interp/ByteCodeExprGen.cpp:299]: Why the different init for first and second element? Complex is basiclly an array underneath: https://eel.is/c++draft/complex.numbers#general-4 but I am not sure I understand the difference in how we init the first and second element. 

Seems like a comment would be helpful.

Comment (tbaederr) [reply clang/lib/AST/Interp/ByteCodeExprGen.cpp:299]: That's just how it works, i.e. in https://github.com/llvm/llvm-project/pull/75485/ we return `0` for `__imag(primitive)`, so we initialize the imaginary part to `0` here when creating a complex number from a primitive.

Comment (tbaederr) [reply clang/lib/AST/Interp/ByteCodeExprGen.cpp:299]: Is that clear enough or should I add a comment?

**Comment by tbaederr** [comment]
Ping

**Comment by tbaederr** [comment]
Ping

**Comment by tbaederr** [comment]
Ping @shafik

File: clang/lib/AST/Interp/ByteCodeExprGen.cpp (hunk 2 above)
Comment (AaronBallman) [reply clang/lib/AST/Interp/ByteCodeExprGen.cpp:299]: Might as well add a comment as to why we zero init.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Pull request merged**